### PR TITLE
[httpx][Makefile][copier] add --timeout to copier; use in Makefile

### PR DESCRIPTION
**Pull request by danking**
We have long had trouble with copying on flaky Internet connections. AFAIK, this is due to our very aggressive timeouts. This change permits the CLI to override the default timeout in `hailtop.aiotools.copy`. Setting this back to 600s should work on most flaky Internet connections. It works on my particularly bad home WiFi.

The upload-docs target was old and broken so I deleted it. I did not change the upload-artifacts target because that is Google Cloud specific anyway.

Main benefit is that these targets work in Azure now.

### CODE REVIEW

**Comment by danking** [comment]
Sent this to you, Ed, since, IIRC, you experienced broken uploads in the past due to our bad copy implementation. If you could clone this and try
```
rm hail/upload-remote-test-resources && make -C hail upload-remote-test-resources
```
that'd be most appreciated!

**Comment by ehigham** [comment]
From a fresh clone, the above (modified with `rm -f`) fails with: 
```bash
$ rm -f hail/upload-remote-test-resources && make -C hail upload-remote-test-resources
make: Entering directory '/home/edmund/.local/src/hail/hail'
# # If hailtop.aiotools.copy gives you trouble:
# gcloud storage cp -r src/test/resources/\* gs://hail-test-ezlis/edmund/hail-test-resources/test/resources/
# gcloud storage cp -r python/hail/docs/data/\* gs://hail-test-ezlis/edmund/hail-test-resources/doctest/data/
python3 -m hailtop.aiotools.copy -vvv 'null' '[\
             {"from":"src/test/resources","to":"gs://hail-test-ezlis/edmund/hail-test-resources/test/resources/"},\
             {"from":"python/hail/docs/data","to":"gs://hail-test-ezlis/edmund/hail-test-resources/doctest/data/"}\
        ]' --timeout 600
/home/edmund/.local/src/hail/.venv/bin/python3: Error while finding module specification for 'hailtop.aiotools.copy' (ModuleNotFoundError: No module named 'hailtop')
make: *** [Makefile:355: upload-remote-test-resources] Error 1
make: Leaving directory '/home/edmund/.local/src/hail/hail'
```

I'll try again with `hailtop` installed - just wanted to point out the dependency failure in `Makefile`.

**Comment by ehigham** [comment]
Likely a kubectl config. Sigh.

**Comment by danking** [comment]
@ehigham `make --version`? This is me:
```
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
```

**Comment by danking** [comment]
In particular, it appears that your Make might be invoking bash with the sixth argument starting with the three characters `'[\`. My Make sends the string `'[ ` because it sees the `\` as a *make* line-continuation character. In particular, the full bash command that my Make invokes is:

```
python3 -m hailtop.aiotools.copy -vvv 'null' [             {"from":"src/test/resources","to":"gs://hail-test-ezlis/dking/hail-test-resources/test/resources/"},             {"from":"python/hail/docs/data","to":"gs://hail-test-ezlis/dking/hail-test-resources/doctest/data/"}        ]
```

**Comment by danking** [comment]
e.g.
```
# cat Makefile
foo:
	echo '[ \
hello \
'
# make foo
echo '[ \
hello \
'
[ hello 
```

If you create such a Makefile and run Make do you see `[ hello ` echo'd or do you see something else, possibly including newlines?

**Comment by danking** [comment]
Good catch on the dependency issue, that should be fixed now.

**Comment by ehigham** [comment]
Make version:
```
$ make --version    
GNU Make 4.4.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by ehigham** [comment]
According to the makefile documentation https://www.gnu.org/software/make/manual/html_node/Splitting-Recipe-Lines.html:
```
Notice how the backslash/newline pair was removed inside the string quoted with double quotes ("…"), but not from the string quoted with single quotes ('…'). This is the way the default shell (/bin/sh) handles backslash/newline pairs. If you specify a different shell in your makefiles it may treat them differently.
```

Seems you (or `brew`) may have configured `make` to use something other than `/bin/sh`.

File: hail/Makefile [hunk 1]
```diff
@@ -347,15 +344,16 @@ upload-artifacts: $(WHEEL)
...
-        #      {"from":"src/test/resources","to":"$(CLOUD_HAIL_TEST_RESOURCES_DIR)"},\
-        #      {"from":"python/hail/docs/data","to":"$(CLOUD_HAIL_DOCTEST_DATA_DIR)"}\
-        # ]'
+	# # If hailtop.aiotools.copy gives you trouble:
+	# gcloud storage cp -r src/test/resources/\* $(CLOUD_HAIL_TEST_RESOURCES_DIR)
+	# gcloud storage cp -r python/hail/docs/data/\* $(CLOUD_HAIL_DOCTEST_DATA_DIR)
+	python3 -m hailtop.aiotools.copy -vvv 'null' '[\
+             {"from":"src/test/resources","to":"$(CLOUD_HAIL_TEST_RESOURCES_DIR)"},\
+             {"from":"python/hail/docs/data","to":"$(CLOUD_HAIL_DOCTEST_DATA_DIR)"}\
+        ]' --timeout 600
```
Comment (ehigham) [inline hail/Makefile:356]:
```suggestion
	python3 -m hailtop.aiotools.copy -vvv 'null' "[\
             {\"from\":\"src/test/resources\",\"to\":\"$(CLOUD_HAIL_TEST_RESOURCES_DIR)\"},\
             {\"from\":\"python/hail/docs/data\",\"to\":\"$(CLOUD_HAIL_DOCTEST_DATA_DIR)\"}\
        ]" --timeout 600
	touch $@
```

**Comment by danking** [comment]
I have bash 3.2.57
```
(base) dking@wm28c-761 /tmp % make print-shell
/bin/sh
(base) dking@wm28c-761 /tmp % /bin/sh --version
GNU bash, version 3.2.57(1)-release (arm64-apple-darwin22)
Copyright (C) 2007 Free Software Foundation, Inc.
```

Looks like this was an intentionally backwards incompatible change [in Make 4.0](https://git.savannah.gnu.org/cgit/make.git/tree/NEWS?h=4.0&id=52191d9d613819a77a321ad6c3ab16e1bc73c381#n18) which removed the POSIX-compatible behavior on which our Makefile relies:
```
* WARNING: Backward-incompatibility!
  If .POSIX is specified, then make adheres to the POSIX backslash/newline
  handling requirements, which introduces the following changes to the
  standard backslash/newline handling in non-recipe lines:
  * Any trailing space before the backslash is preserved
  * Each backslash/newline (plus subsequent whitespace) is converted to a
    single space
```

They seem to have [broken the behavior in order to fix something else](https://git.savannah.gnu.org/cgit/make.git/commit/?id=391456a) and then added a [`.POSIX`](https://git.savannah.gnu.org/cgit/make.git/commit/?id=88f1bc8) escape hatch for Makefiles that want POSIX compatibility.

For the Makefile I shared above, I get equivalent behavior with and without a `.POSIX:` fake target. I suspect Make 4.x treats them differently.

In particular, the problem is not the shell, it's our Makes behaving differently. Make 3 replaces `\\\n` with ` ` before calling the shell whereas Make 4 appears to preserve the `\\\n` to the shell (regardless of the presence of quotes in either case). In particular, Make 4 appears to treat recipe lines differently from how it treats all other lines in Makefiles.

I don't like the noise of the backslash-quotes. The end of the doc page you linked suggests using a Make variable when you intentionally want normal Makefile backslash-newline interpretation. That feels a bit too clever. Let's try just demanding POSIX-compatibility. I've pushed that change. Can you give it a try?

**Comment by ehigham** [comment]
You're absolutely right - my `/bin/sh` is also a symlink to bash!

**Review by ehigham — APPROVED** [review]
Works like a charm

**Pull request merged**